src/msw/printwin.cpp

```diff
@@ -163,16 +163,40 @@ bool wxWindowsPrinter::Print(wxWindow *parent, wxPrintout *printout, bool prompt
     sm_lastError = wxPRINTER_NO_ERROR;
 
     int minPageNum = minPage, maxPageNum = maxPage;
+    int totalPageCnt = -1;
 
     if ( m_printDialogData.GetCurrentPage() || m_printDialogData.GetSelection() )
     {
-        minPageNum = fromPage;
-        maxPageNum = toPage;
+        if (fromPage > 0 && toPage > 0)
+        {
+            minPageNum = fromPage;
+            maxPageNum = toPage;
+        }
+        totalPageCnt = maxPageNum - minPageNum + 1;
     }
     else if ( !m_printDialogData.GetAllPages() )
     {
-        minPageNum = m_printDialogData.GetFromPage();
-        maxPageNum = m_printDialogData.GetToPage();
+        if (m_printDialogData.GetFromPage() > 0 && m_printDialogData.GetToPage() > 0)
```

Comment:
It's not really clear how does `Get{From,To}Page()` and `GetPageRanges()` work together. Can we have both of them valid at the same time? It looks like the former should be just a special case of the latter, with a single range, could we do it like this or is there something that prevents us from doing it?

Reply:
Can be used together and is there because this is backward compatible. The ranges array is currently just an extension. If you want to use more than one range, you have to use the ranges array.

Reply:
`m_printFromPage` and `m_printToPage` are private, so it looks like we could get rid of them and just use the first element of `m_printPageRanges` to implement `{Get,Set}{From,To}Page()` instead, couldn't we?

`GetPageInfo()` is incompatible with using multiple ranges, so we need to extend the API and add some `vector<wxPrintPageRange> GetPageRanges() const` which would be called by wx. The default implementation would call `GetPageInfo()` for compatibility, but it could be overridden to return richer information, with multiple ranges.

This is not going to be that simple, but AFAICS this problem can be solved and I'm almost sure that it has to be solved to allow the application to specify multiple range of pages to print as this still seems to be unsupported with the current version of the PR.

But maybe we need a deeper rethink of how this stuff works because it seems suspicious to have similar but not at all the same information in `wxPrintDialogData` and `GetPageInfo()`... Maybe we should just pass the former to the latter and allow it to modify it? Sorry, I'm running out of time here :-(

Reply:
I have revised the code and made the changes you suggested. Please have a look. I've also added some more comments to make some things a little more clear.